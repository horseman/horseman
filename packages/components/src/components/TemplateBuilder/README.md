# TemplateBuilder
The template builder is intended to render out a series of components from json
data. It's primary use case it to build dynamic pages from data returned from a
CMS api.

## Example
```js
import { TemplateBuilder } from "horseman-components";

import Content from "components/Content";
import Header from "components/Header";
import Picture from "components/Picture";
import Carousel from "components/Carousel";

// Map named keys to components to render.
const mappings = {
  content: Content,
  image: Picture,
  carousel: Carousel,
  header: Header,
};

// Most likely this data should originate from an external API.
const data = [
  {
    type: "content",
    body: "Dolor officiis non officiis tenetur debitis eveniet. Hic praesentium dolor."
  },
  {
    type: "carousel",
    images: [
      {
        src: "http://placehold.it/100x100",
      },
      {
        src: "http://placehold.it/200x200",
      },
    ],
  },
  {
    type: "header",
    text: "Foo Bar"
  },
  {
    type: "header",
    text: "Foo Bar"
  },
];


// Render TemplateBuilder component
<TemplateBuilder mapings={mappings}, data={data} />
```
