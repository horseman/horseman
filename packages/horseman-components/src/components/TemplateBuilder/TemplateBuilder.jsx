import React from "react";
import PropTypes from "prop-types";

const TemplateBuilder = ({ data, mappings }) => {
  const items = data.map((block, i) => {
    const { type, ...rest } = block;
    const Component = mappings[type];

    return typeof Component === "undefined" ? null : (
      // eslint-disable-next-line react/no-array-index-key
      <Component key={i} {...rest} />
    );
  });

  return <div>{items}</div>;
};

TemplateBuilder.defaultProps = {
  data: [],
};

TemplateBuilder.propTypes = {
  /**
   * The data that will be used to generate the templates.
   */
  mappings: PropTypes.object.isRequired,

  /**
   * The page builder array that will generate the pageBuilder.  Will render a
   * page builder block based on the `type` of block requested. Block types are
   * registered in the imported mappings file.
   */
  data: PropTypes.arrayOf(
    PropTypes.shape({
      type: PropTypes.string.isRequired,
    }),
  ),
};

export default TemplateBuilder;
