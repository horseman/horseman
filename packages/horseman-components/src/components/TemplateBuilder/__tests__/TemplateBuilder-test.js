/* eslint-disable react/prop-types */
import React from "react";
import { shallow } from "enzyme";

import TemplateBuilder from "../";

describe("TemplateBuilder", () => {
  const Content = ({ body }) => <div>{body}</div>;
  const Image = ({ src }) => <img alt=" " src={src} />;

  const data = [
    {
      type: "content",
      body: "foo bar",
    },
    {
      type: "image",
      src: "bizbaz",
    },
    {
      type: "undefined",
    },
  ];

  const mappings = {
    content: Content,
    image: Image,
  };

  test("renders correctly", () => {
    const wrapper = shallow(
      <TemplateBuilder data={data} mappings={mappings} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
