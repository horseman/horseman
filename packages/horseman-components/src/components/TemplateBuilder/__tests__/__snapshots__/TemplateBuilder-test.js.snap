// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TemplateBuilder renders correctly 1`] = `
<div>
  <Content
    body="foo bar"
    key="0"
  />
  <Image
    key="1"
    src="bizbaz"
  />
</div>
`;
